sklearn/tests/test_props.py

```diff
@@ -0,0 +1,729 @@
+    clf = SimpleMetaClassifier(
+        estimator=ClassifierFitMetadata().set_fit_request(
+            sample_weight="alternative_weight"
+        )
+    )
+    clf.fit(X, y, alternative_weight=my_weights)
+    check_recorded_metadata(clf.estimator_, "fit", sample_weight=my_weights, brand=None)
+
+
+def test_nested_routing():
+    # check if metadata is routed in a nested routing situation.
+    pipeline = SimplePipeline(
+        [
+            MetaTransformer(
+                transformer=TransformerMetadata()
+                .set_fit_request(brand=True, sample_weight=False)
+                .set_transform_request(sample_weight=True)
+            ),
+            WeightedMetaRegressor(
+                estimator=RegressorMetadata().set_fit_request(
+                    sample_weight="inner_weights"
+                )
+            ).set_fit_request(sample_weight="outer_weights"),
+        ]
+    )
+    w1, w2, w3 = [1], [2], [3]
+    pipeline.fit(
+        X, y, brand=my_groups, sample_weight=w1, outer_weights=w2, inner_weights=w3
+    )
+    check_recorded_metadata(
+        pipeline.steps_[0].transformer_, "fit", brand=my_groups, sample_weight=None
+    )
+    check_recorded_metadata(
+        pipeline.steps_[0].transformer_, "transform", sample_weight=w1
+    )
+    check_recorded_metadata(pipeline.steps_[1], "fit", sample_weight=w2)
+    check_recorded_metadata(pipeline.steps_[1].estimator_, "fit", sample_weight=w3)
+
+    pipeline.predict(X, sample_weight=w3)
+    check_recorded_metadata(
+        pipeline.steps_[0].transformer_, "transform", sample_weight=w3
+    )
+
+
+def test_nested_routing_conflict():
+    # check if an error is raised if there's a conflict between keys
+    pipeline = SimplePipeline(
+        [
+            MetaTransformer(
+                transformer=TransformerMetadata()
+                .set_fit_request(brand=True, sample_weight=False)
+                .set_transform_request(sample_weight=True)
+            ),
+            WeightedMetaRegressor(
+                estimator=RegressorMetadata().set_fit_request(sample_weight=True)
+            ).set_fit_request(sample_weight="outer_weights"),
+        ]
+    )
+    w1, w2 = [1], [2]
+    with pytest.raises(
+        ValueError,
+        match=(
+            re.escape(
+                "In WeightedMetaRegressor, there is a conflict on sample_weight between"
+                " what is requested for this estimator and what is requested by its"
+                " children. You can resolve this conflict by using an alias for the"
+                " child estimator(s) requested metadata."
+            )
+        ),
+    ):
+        pipeline.fit(X, y, brand=my_groups, sample_weight=w1, outer_weights=w2)
+
+
+def test_invalid_metadata():
+    # check that passing wrong metadata raises an error
+    trs = MetaTransformer(
+        transformer=TransformerMetadata().set_transform_request(sample_weight=True)
+    )
+    with pytest.raises(
+        ValueError,
```

Comment:
This should probably be a TypeError, matching the existing behaviour if a method is called with an unknown param name. The traditional error message is `{method} got an unexpected keyword argument {key}`, and we could extend this `{method} got an unexpected keyword argument {key}, which is not requested metadata in any object`